File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# (깊이, 값 )

def solution(numbers, target):
    
    ans = 0
    
    stack = [(0, numbers[0]), (0, -numbers[0])]
    
    
    while stack:
        
        depth, number = stack.pop()
        
        if depth==(len(numbers)-1) and number == target:
            ans += 1
            
        
        if depth < len(numbers)-1:
            stack.append((depth+1, number+numbers[depth+1]))
            stack.append((depth+1, number-numbers[depth+1]))
        
        
    return ans

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <bits/stdc++.h>
using namespace std;

int answer = 0;

void dfs(vector<int> numbers, int target, int idx, int sum) {
    if (idx == numbers.size()) {
        if (sum == target) answer++;
        return;
    }
    // 현재 숫자 더하기
    dfs(numbers, target, idx+1, sum + numbers[idx]);
    // 현재 숫자 빼기
    dfs(numbers, target, idx+1, sum - numbers[idx]);
}

int solution(vector<int> numbers, int target) {
    dfs(numbers, target, 0, 0);
    return answer;
}
